lib/rules/no-useless-escape.js

```diff
@@ -55,7 +55,71 @@ const VALID_REGEX_ESCAPES = [
     "W",
     "x",
     "u"
-];
+]);
+const REGEX_CHARCLASS_ESCAPES = union(REGEX_GENERAL_ESCAPES, new Set(["]"]));
+const REGEX_NON_CHARCLASS_ESCAPES = union(REGEX_GENERAL_ESCAPES, new Set([
+    "/",
+    ".",
+    "$",
+    "*",
+    "+",
+    "?",
+    "[",
+    "{",
+    "}",
+    "|",
+    "(",
+    ")",
+]));
+
+/**
+* Parses a regular expression into a list of characters with character class info.
+* @param {string} regExpText The raw text used to create the regular expression
+* @returns {Object[]} A list of characters, each with info on escaping and whether they're in a character class.
+* @example
+*
+* parseRegExp('a\\b[cd-]')
+*
+* returns:
+* [
+*   {text: 'a', index: 0, escaped: false, charClass: false},
+*   {text: 'b', index: 2, escaped: true, charClass: false},
+*   {text: 'c', index: 4, escaped: false, charClass: true},
+*   {text: 'd', index: 5, escaped: false, charClass: true},
+*   {text: '-', index: 6, escaped: false, charClass: true}
+* ]
+*/
+function parseRegExp(regExpText) {
+    return regExpText.split("").reduce((state, char, index) => {
+        if (!state.escapeNextChar) {
+            if (char === "\\") {
+                return Object.assign({}, state, {escapeNextChar: true, index});
+            }
+            if (char === "[" && !state.inCharClass) {
+                return Object.assign({}, state, {inCharClass: true, index});
+            }
+            if (char === "]" && state.inCharClass) {
+                return Object.assign({}, state, {inCharClass: false, index});
+            }
+        }
+        return {
+            charList: state.charList.concat({text: char, escaped: state.escapeNextChar, inCharClass: state.inCharClass, index}),
```

Comment:
Does it need to create an array instance for each character? It looks to generate many useless copies.